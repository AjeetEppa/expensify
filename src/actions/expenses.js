import { v4 as uuid } from "uuid";
import * as database from "firebase/database";

// Expense Actions Generators

export const addExpense = (expense) => ({
  type: "ADD_EXPENSE",
  expense,
});

export const startAddExpense = (expenseData = {}) => {
  const {
    description = "",
    note = "",
    amount = 0,
    createdAt = 0,
  } = expenseData;
  const expense = { description, note, amount, createdAt };
  return (dispatch) => {
    const db = database.getDatabase();
    return database.push(database.ref(db, "expenses"), expense).then((ref) => {
      dispatch(addExpense({ id: ref.key, ...expense }));
    });
  };
};

export const removeExpense = ({ id = "" } = {}) => ({
  type: "REMOVE_EXPENSE",
  id,
});

export const editExpense = (id, updates) => ({
  type: "EDIT_EXPENSE",
  id,
  updates,
});
