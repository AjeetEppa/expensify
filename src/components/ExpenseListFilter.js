import React from "react";
import { Component } from "react";
import { DateRangePicker } from "react-dates";
import { connect } from "react-redux";
import {
  setEndDate,
  setStartDate,
  setTextFilter,
  sortByAmount,
  sortByDate,
} from "../actions/filters";

export class ExpenseListFilter extends Component {
  state = {
    calendarFocused: null,
  };
  onDatesChange = ({ startDate, endDate }) => {
    console.log(startDate, endDate);
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };
  onCalendarFocusChange = (focus) => {
    this.setState({ calendarFocused: focus });
  };
  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
  };
  onSortByChange = (e) => {
    switch (e.target.value) {
      case "amount":
        this.props.sortByAmount();
        break;
      case "date":
        this.props.sortByDate();
        break;
      default:
        break;
    }
  };
  render() {
    return (
      <div>
        <input
          type="text"
          value={this.props.filters.text}
          onChange={this.onTextChange}
        />
        <select
          value={this.props.filters.sortBy}
          onChange={this.onSortByChange}
        >
          <option value="amount">Amount</option>
          <option value="date">Date</option>
        </select>
        <DateRangePicker
          startDate={this.props.filters.startDate} // momentPropTypes.momentObj or null,
          startDateId="your_unique_start_date_id" // PropTypes.string.isRequired,
          endDate={this.props.filters.endDate} // momentPropTypes.momentObj or null,
          endDateId="your_unique_end_date_id" // PropTypes.string.isRequired,
          onDatesChange={this.onDatesChange} // PropTypes.func.isRequired,
          focusedInput={this.state.calendarFocused} // PropTypes.oneOf([START_DATE, END_DATE]) or null,
          onFocusChange={this.onCalendarFocusChange} // PropTypes.func.isRequired,
          numberOfMonths={1}
          isOutsideRange={(date) => false}
          showClearDates
        />
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  filters: state.filters,
});

const mapDispatchToProps = (dispach) => ({
  setStartDate: (date) => dispach(setStartDate(date)),
  setEndDate: (date) => dispach(setEndDate(date)),
  setTextFilter: (text) => dispach(setTextFilter(text)),
  sortByAmount: () => dispach(sortByAmount()),
  sortByDate: () => dispach(sortByDate()),
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilter);
