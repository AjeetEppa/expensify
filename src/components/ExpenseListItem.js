import React from "react";
import { Link } from "react-router-dom";

const ExpenseListItem = (props) => {
  return (
    <>
      <Link to={`/edit/${props.id}`}>
        <h2>{props.description}</h2>
      </Link>
      <p>
        Amount: {props.amount} | Create at: {props.createdAt}
      </p>
    </>
  );
};

export default ExpenseListItem;
