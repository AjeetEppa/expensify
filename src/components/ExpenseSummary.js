import numeral from "numeral";
import React from "react";
import { connect } from "react-redux";
import selectExpensesTotal from "../selectors/expenses-total";
import selectExpenses from "../selectors/expenses";

export const ExpenseSummary = (props) => {
  const { expenseCount, expenseTotal } = props;
  return (
    <h3>
      Viewing {expenseCount || "0"} expense{expenseCount == 1 ? "" : "s"}{" "}
      totalling {numeral(expenseTotal / 100).format("$0,0.00")}
    </h3>
  );
};

const mapStateToProps = (state) => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);
  console.log(state);
  return {
    expenseCount: visibleExpenses.length,
    expenseTotal: selectExpensesTotal(visibleExpenses),
  };
};

export default connect(mapStateToProps)(ExpenseSummary);
