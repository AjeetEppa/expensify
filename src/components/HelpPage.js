import React from "react";

const HelpPage = () => <p>This is help component</p>;

export default HelpPage;
