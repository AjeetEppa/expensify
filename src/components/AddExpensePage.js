import React from "react";
import { connect } from "react-redux";
import { addExpense } from "../actions/expenses";
import ExpenseForm from "./ExpenseForm";

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.addExpense(expense);
    this.props.history.push("/");
  };
  render() {
    return <ExpenseForm onSubmit={this.onSubmit} />;
  }
}

const mapDispatchToProps = (dispatch) => ({
  addExpense: (expense) => {
    dispatch(addExpense(expense));
  },
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
