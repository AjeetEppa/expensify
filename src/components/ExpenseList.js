import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from "../selectors/expenses";

export const ExpenseList = (props) => {
  return !props.expenses.length ? (
    <p>No Expenses</p>
  ) : (
    <div>
      {props.expenses.map((prop, i) => (
        <ExpenseListItem key={i} {...prop} />
      ))}
    </div>
  );
};

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters),
  };
};

export default connect(mapStateToProps)(ExpenseList);
