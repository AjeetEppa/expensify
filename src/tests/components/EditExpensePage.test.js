import React from "react";
import { shallow } from "enzyme";
import { EditExpensePage } from "../../components/EditExpensePage";
import ExpenseForm from "../../components/ExpenseForm";
import expenses from "../fixtures/expenses";

let expense, editExpense, removeExpense, history, wrapper;

beforeEach(() => {
  editExpense = jest.fn();
  removeExpense = jest.fn();
  history = { push: jest.fn() };
  expense = expenses[0];
  wrapper = shallow(
    <EditExpensePage
      expense={expense}
      editExpense={editExpense}
      removeExpense={removeExpense}
      history={history}
    />
  );
});

test("should render EditExpensePage correctly", () => {
  expect(wrapper).toMatchSnapshot();
});

test("should handle editExpense", () => {
  wrapper.find(ExpenseForm).prop("onSubmit")(expense);
  expect(history.push).toHaveBeenCalledWith("/");
  expect(editExpense).toHaveBeenCalledWith(expense.id, expense);
});

test("should render removeExpense", () => {
  wrapper.find("button").simulate("click");
  expect(history.push).toHaveBeenCalledWith("/");
  expect(removeExpense).toHaveBeenCalledWith({ id: expense.id });
});
