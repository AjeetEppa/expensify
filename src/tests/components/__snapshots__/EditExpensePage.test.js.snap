// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage correctly 1`] = `
<Fragment>
  <ExpenseForm
    expense={
      Object {
        "amount": 2500,
        "createdAt": 0,
        "description": "Hathway",
        "id": "1",
        "note": "Intenet connection",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</Fragment>
`;
